58363 Automatically disable ANSI escape sequences in Pretty formats if the output is not a terminal

Blargian opened the pull request:
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Automatically disable ANSI escape sequences in Pretty formats if the output is not a terminal. Add new `auto` mode to setting `output_format_pretty_color`. 

Hi, 

I'm opening a draft pull request for #58363 (hope that's okay!). I implemented the following changes so far together with a stateless test to check the functionality.

**Changes made:**
- modified `output_format_pretty_color` to use options `Off`, `On`, `Auto` (`Core/Settings.h` and `FormatFactory.cpp`)
- Added an `enum class` `PrettyColor` with these options (`FormatSettings.h)`
- Modified `PrettyBlockOutputFormat.h` to switch on the settings above and turn `pretty.color` on or off based on conditions provided in the issue.
- Added `00405_output_format_pretty_color.sql` and generated the reference file. 

**Help points:**

-  The behaviour appears correct to me for `On` and `Off` but I don't believe I have the correct usage of `isatty` function for the `Auto` setting. I'm unsure on this point. 
- I went with `On`,`Off` and `Auto` but you did mention that you want `0`,`1` and `auto` as the setting options. Could I achieve this somehow with the `enum class` or is there a better alternative?
- Any comments on formatting / performance
- Manually testing from `http://localhost:8123/play` the behaviour is not correct. It is different to the reference file, so perhaps it is user trouble.

Blargian committed:
#58363 :construction: modified Pretty in FormatSettings.h to have PrettyColor which can be 0,1 or auto. modified output_format_pretty_color in FormatFactory.cpp to make use of this, added the default to Settings.h. Implemented the logic for enabling/disabling based on output_format_pretty_color in PrettyBlockOutputFormat.h

Blargian committed:
#58363 - fix formatting issues and change ON, OFF, AUTO to 0, 1, auto

Blargian committed:
minor fixes. Doesnt seem to be using ANSI escapes anymore

Blargian committed:
#58363 - added stateless test with expected behaviour for Off and On settings but not Auto

robot-clickhouse-ci-1 commented:
<!-- automatic status comment for PR #58614 from Blargian/ClickHouse:58363_disable_ansi_pretty_automatically -->
*This is an automated comment for commit 76ba291c9d37ce0ab184552a337e2523efbabdb7 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58614/76ba291c9d37ce0ab184552a337e2523efbabdb7/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

Avogar commented:
> I went with On,Off and Auto but you did mention that you want 0,1 and auto as the setting options. Could I achieve this somehow with the enum class or is there a better alternative?

You can use special struct-wrapper `UInt64Auto` for it. See as an example setting `insert_quorum`:
https://github.com/ClickHouse/ClickHouse/blob/dbb19e72b38f9ee8848b4fcc15ecb007eb929751/src/Core/Settings.h#L267

Avogar commented on src/Processors/Formats/Impl/PrettyBlockOutputFormat.h:
```diff
@@ -73,10 +73,22 @@ void registerPrettyFormatWithNoEscapesAndMonoBlock(FormatFactory & factory, cons
+                    changed_settings.pretty.color = no_escapes ? false : true; 
+                    break;
+                case PrettyColor::Auto:
+                    changed_settings.pretty.color = isatty(STDOUT_FILENO) && no_escapes ? false : true; 
```
We will output data into our `WriteBuffer`, not always to stdout. So, we should first check if our buffer will output data into stdout and then use `isatty` function on `STDOUT_FILENO`. We can create a separate function for it:
```
bool isWritingToTerminal(const WriteBuffer & buf)
{
    const auto * write_buffer_to_descriptor = typeid_cast<const WriteBufferFromFileDescriptor *>(&buf);
    return write_buffer_to_descriptor && write_buffer_to_descriptor->getFD() == STDOUT_FILENO && isatty(STDOUT_FILENO);
}
```

We can place this function in file `src/IO/WriteHelpers.h`.

Also if I understood correctly, we should do `|| no_escapes` here, not `&& no_escapes`. So it will look like this:
```
                    changed_settings.pretty.color = isWritingToTerminal(buf) || no_escapes ? false : true; 
```

Blargian committed:
#58363 - Changes based on review of draft PR - changed output_format_pretty_color to use UInt64Auto. Added isWritingToTerminal function to IO/WriteHelpers.h and updated test

Blargian commented:
Hi @Avogar, I've implemented the changes you mentioned in the latest commit. 
- Using `UInt64Auto` now for `output_format_pretty_color` setting
- Added  `isWritingToTerminal` to `IO/WriteHelpers.h` (I marked it `inline` otherwise we get a "multiple defined symbols linker error")
- Updated the test to reflect the changes and the reference file seems to display the behaviour we are looking for now

One thing which is still bothering me, is that if I manually test using `http://localhost:8123/play` and run 
```sql
SET output_format_pretty_color = 0
SELECT number AS hello, toString(number) AS world, (hello, world) AS tuple, nullIf(hello % 3, 0) AS sometimes_nulls FROM system.numbers LIMIT 10 SETTINGS max_block_size = 5 FORMAT Pretty;
```
I get back an output with ANSI Escapes still. But maybe I'm just not changing the setting correctly because if I run query 

```
SELECT * FROM system.settings WHERE name='output_format_pretty_color'
```

value is still `auto`?

Avogar commented on src/Formats/FormatSettings.h:
```diff
@@ -270,6 +270,8 @@ struct FormatSettings
         UInt64 max_value_width = 10000;
         bool color = true;
+	
+		SettingFieldUInt64Auto output_format_pretty_color{"auto"};
```
No need to add a new setting for it. We can just change the type of `color` setting.

Avogar commented on src/Processors/Formats/Impl/PrettyBlockOutputFormat.h:
```diff
@@ -73,10 +73,22 @@ void registerPrettyFormatWithNoEscapesAndMonoBlock(FormatFactory & factory, cons
             const FormatSettings & format_settings)
         {
-            if (no_escapes)
+            FormatSettings changed_settings = format_settings;
+			auto value = format_settings.pretty.output_format_pretty_color.valueOr(2);
+            switch (value)
+            {
+                case 0:
+                    changed_settings.pretty.color = false;
+                    break;
+                case 1:
+                    changed_settings.pretty.color = no_escapes ? false : true; 
+                    break;
+                case 2:
+                    changed_settings.pretty.color = isWritingToTerminal(buf) ||  no_escapes ? false : true; 
+                    break;
+            } 
+            if (!changed_settings.pretty.color)
```
It will be better not copy and change `format_settings`. We can add a bool flag `color` to classes `PrettyBlockOutputFormat/PrettyBlockOutputFormat` constructors and use this flag instead of value from settings. So, the code here will look like this:
```
bool color = !no_escapes && format_settings.pretty.color.valueOr(isWritingToTerminal(buf));
return std::make_shared<OutputFormat>(buf, sample, format_settings, mono_block, color);
```

Blargian replied:
Agreed, copy of `format_settings` seems unnecessary. Would we not have to change `PrettyCompactBlockOutputFormat` and `PrettySpaceBlockOutputFormat` constructors as well in this case, not only `PrettyBlockOutputFormat`?

Avogar replied:
Yes, all `PrettyBlockOutputFormat/PrettyCompactBlockOutputFormat/PrettySpaceBlockOutputFormat`. Should not be difficult.

I just noticed that in the comment I just wrote duplicated `PrettyBlockOutputFormat/PrettyBlockOutputFormat `, I meant all these three actually :)

Blargian replied:
I thought that was what you meant, just checking )

Blargian replied:
@Avogar adding `color` to the constructor is easy enough, my only concern with this approach is that we mess with the consistency of the classes a little bit. For instance:

```c++
void PrettyCompactBlockOutputFormat::writeRow(
    size_t row_num,
    const Block & header,
    const Columns & columns,
    const WidthsPerColumn & widths,
    const Widths & max_widths)
{
    if (format_settings.pretty.output_format_pretty_row_numbers)
    {
        // Write row number;
        auto row_num_string = std::to_string(row_num + 1 + total_rows) + ". ";
        for (size_t i = 0; i < row_number_width - row_num_string.size(); ++i)
        {   
            writeCString(" ", out);
        }
        writeString(row_num_string, out);
    }
```
We check `output_format_pretty_row_numbers` and `max_rows` on `format_settings` object but we will check `color` as a member variable of `PrettyBlockOutputFormat`. From code cleanliness point of view probably it would be better to be consistent. 

Mind you, this inconsistency already exists in `PrettyBlockOutputFormat.cpp`, some properties we check on `format_settings` but `mono_block` is a member variable by itself. 

What do you think?

Avogar replied:
It's totally ok to use some of the flags from `format_settings` and some from constructor. We already use it in sevaral places in formats.

Avogar commented:
We need to change documentation for this setting in [docs/en/operations/settings/settings-formats.md](https://github.com/ClickHouse/ClickHouse/blob/master/docs/en/operations/settings/settings-formats.md#output_format_pretty_color-output_format_pretty_color)

Avogar commented:
> One thing which is still bothering me, is that if I manually test using http://localhost:8123/play and run <...> I get back an output with ANSI Escapes still but expect not to see any escapes. But maybe I'm just not changing the setting correctly

AFAIK multi-statemens are not allowed in play UI, so we cannot do
```sql
set output_format_pretty_color =...;
select ... format Pretty;
```

And if you just first execute `SET` query and then `SELECT`, then setting change won't take any effect, because each HTTP query has it's own query context with it's own set of settings. To test it you can do:
```sql
select ... format Pretty settings output_format_pretty_color=...;
```

Blargian commented:
> We need to change documentation for this setting in [docs/en/operations/settings/settings-formats.md](https://github.com/ClickHouse/ClickHouse/blob/master/docs/en/operations/settings/settings-formats.md#output_format_pretty_color-output_format_pretty_color)

Are only En language docs kept up to date fully? I notice that Ru documentation contains only some of `output_format_pretty` settings. Можно добавить `output_format_pretty_color` к `docs/ru/operations/settings/settings.md`? ... на сломанном русском... грамматическое качество не гарантирую!

Avogar commented:
> Are only En language docs kept up to date fully?

Yes, ClickHouse team keeps English docs up to date, but other languages are supported as much as possible by external contributors from different countries.

> Можно добавить output_format_pretty_color к docs/ru/operations/settings/settings.md? ... на сломанном русском... грамматическое качество не гарантирую!

Only if you want, it's not necessary. But if you do it, I will check the the grammar

Blargian committed:
#58363 - removed switch from PrettyBlockOutputFormat and modified BlockOutputFormats to use color variable. Updated english and russian documentation. Updated test 00405 reference file.

Blargian commented:
@Avogar I pushed the latest changes. When generating the reference file again on these changes ANSI escapes are no longer shown for `auto` when they should be. If I step through the code with lldb when `outcome_format_pretty_color=auto` the logic seems correct to me, I don't spot something unusual. Eg) for normal `Pretty` format `color` below evaluates as `1` and if I step through in `PrettyBlockOutputFormat.cpp` I can see that color is used correctly. Can you spot anything unusual in the changes I made that might cause the output to be incorrect? 

```c++
bool color = !no_escapes && format_settings.pretty.color.valueOr(isWritingToTerminal(buf));
return std::make_shared<OutputFormat>(buf, sample, format_settings, mono_block, color);
```

Avogar commented:
Ah, the problem is that Pretty format supports parallel formatting and it's enabled by default. During parallel formatting we output blocks of data in Pretty format in parallel in special memory chunks and then output these chunks to the initial write buffer in corresponding order. So when we create PrettyBlockOutputFormat we have WriteBuffer that writes to some memory, not to stdout, `ParallelFormattingOutputFormat` actually works with initial WriteBuffer to stdout.
Probably the best solution will be to add a new common field in `FormatSettings` with name `is_writing_to_terminal` and set it when we create output format in `FormatFactory::getOutputFormat/getOutputFormatParallelIfPossible` when we initially create format settings and have initial WriteBuffer.
Here:
https://github.com/ClickHouse/ClickHouse/blob/43b31204a43f1a9d1c8833ea6e718a865cc9ab79/src/Formats/FormatFactory.cpp#L491-L492
And here:
https://github.com/ClickHouse/ClickHouse/blob/43b31204a43f1a9d1c8833ea6e718a865cc9ab79/src/Formats/FormatFactory.cpp#L521

This new field `is_writing_to_terminal` will be just a common property that any format can use.

Blargian committed:
#58363 - fix failing style check

Blargian committed:
#58363 - fix style check failing attempt no.2

Blargian commented:
@Avogar if I understood correctly, we will set the value of this common field `is_writing_to_terminal` using the helper function `isWritingToTerminal(buf)` previously added to `IO/WriteHelpers`? The idea being that we need to check if we are writing to a terminal *before* the creator function in `PrettyBlockOutputFormat.h` gets called? I.e before we get the format.

Avogar commented:
> if I understood correctly, we will set the value of this common field is_writing_to_terminal using the helper function isWritingToTerminal(buf) previously added to IO/WriteHelpers?

Yes.

> The idea being that we need to check if we are writing to a terminal before the creator function in PrettyBlockOutputFormat.h gets called? I.e before we get the format

Yes, because in creator it's not guaranteed that the write buffer will write to the actual destination, but in `FormatFactory::getOutputFormat/getOutputFormatParallelIfPossible` it's guaranteed.

> and yet another question. I'm trying to understand why I get a different result when running than from when I store the output in the reference file with

I see that the result is expected in both cases. In first case with output to the terminal we have bold column names constructed by escapes as expected. In the second case with output to file I see escape sequances in the result, the only difference that in the second case escape sequances are not displayed properly (probably you used `less` to see the file or smth similar).

Blargian committed:
#58363 - added setting is_writing_to_terminal to FormatSettings.h, modified PrettyBlockOutputFormat to use this, which is set in FormatFactory.cpp getOutputFormat and getOutputFormatParallelIfPossible

Avogar commented:
Many tests failed, but it's expected as we changed the value of the setting. We need to update these tests, either update reference files or set `output_format_pretty_color=1` explicitly in these tests

Avogar commented on src/Core/Settings.h:
```diff
@@ -1041,7 +1041,7 @@ class IColumn;
     M(UInt64, output_format_pretty_max_column_pad_width, 250, "Maximum width to pad all values in a column in Pretty formats.", 0) \
     M(UInt64, output_format_pretty_max_value_width, 10000, "Maximum width of value to display in Pretty formats. If greater - it will be cut.", 0) \
-    M(Bool, output_format_pretty_color, true, "Use ANSI escape sequences to paint colors in Pretty formats", 0) \
+    M(UInt64Auto, output_format_pretty_color, "auto", "Use ANSI escape sequences in Pretty formats. 0 - disabled, 1 - enabled, 'auto' - enabled if a terminal.", 0) \
```
Let's add this change to `SettingsChangesHistory.h`

Blargian commented:
@Avogar okay, clear. I was using vim to view the reference file which is why the escapes `^[[0m` show up like that. Even so something is still wrong. In the reference file with `output_format_pretty_color = auto` and using `FORMAT Pretty` we see this:

```
┏━━━━━━━┳━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ hello ┃ world ┃ tuple   ┃ sometimes_nulls ┃
┡━━━━━━━╇━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│     0 │ 0     │ (0,'0') │            ᴺᵁᴸᴸ │
├───────┼───────┼─────────┼─────────────────┤
│     1 │ 1     │ (1,'1') │               1 │
├───────┼───────┼─────────┼─────────────────┤
│     2 │ 2     │ (2,'2') │               2 │
├───────┼───────┼─────────┼─────────────────┤
│     3 │ 3     │ (3,'3') │            ᴺᵁᴸᴸ │
├───────┼───────┼─────────┼─────────────────┤
│     4 │ 4     │ (4,'4') │               1 │
└───────┴───────┴─────────┴─────────────────┘
``` 

and for `FORMAT PrettyNoEscapes` we see this:

```
┏━━━━━━━┳━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ hello ┃ world ┃ tuple   ┃ sometimes_nulls ┃
┡━━━━━━━╇━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│     0 │ 0     │ (0,'0') │            ᴺᵁᴸᴸ │
├───────┼───────┼─────────┼─────────────────┤
│     1 │ 1     │ (1,'1') │               1 │
├───────┼───────┼─────────┼─────────────────┤
│     2 │ 2     │ (2,'2') │               2 │
├───────┼───────┼─────────┼─────────────────┤
│     3 │ 3     │ (3,'3') │            ᴺᵁᴸᴸ │
├───────┼───────┼─────────┼─────────────────┤
│     4 │ 4     │ (4,'4') │               1 │
└───────┴───────┴─────────┴─────────────────┘
```

However, we should actually see for `FORMAT Pretty` this, as we see when `output_format_pretty_color = 1`:

```
┏━━━━━━━┳━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ [1mhello[0m ┃ [1mworld[0m ┃ [1mtuple  [0m ┃ [1msometimes_nulls[0m ┃
┡━━━━━━━╇━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│     0 │ 0     │ (0,'0') │            ᴺᵁᴸᴸ │
├───────┼───────┼─────────┼─────────────────┤
│     1 │ 1     │ (1,'1') │               1 │
├───────┼───────┼─────────┼─────────────────┤
│     2 │ 2     │ (2,'2') │               2 │
├───────┼───────┼─────────┼─────────────────┤
│     3 │ 3     │ (3,'3') │            ᴺᵁᴸᴸ │
├───────┼───────┼─────────┼─────────────────┤
│     4 │ 4     │ (4,'4') │               1 │
└───────┴───────┴─────────┴─────────────────┘
```

CI is also failing on Fast tests for tests which are using `Pretty` format. 

What is confusing me is that if I step through the code for `Pretty` and `PrettyNoEscapes`, `color` is getting set as we would expect - `true` for `Pretty` but `false` for `PrettyNoEscapes`. 

As you mentioned we should see the bold column names for `Pretty` however I get the exact same formatting output to the terminal for both `Pretty` and `PrettyNoEscapes`. 

Running `SELECT number AS hello, toString(number) AS world, (hello, world) AS tuple, nullIf(hello % 3, 0) AS sometimes_nulls FROM system.numbers LIMIT 10 SETTINGS max_block_size = 5 FORMAT Pretty;`

```
┏━━━━━━━┳━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ hello ┃ world ┃ tuple   ┃ sometimes_nulls ┃
┡━━━━━━━╇━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│     0 │ 0     │ (0,'0') │            ᴺᵁᴸᴸ │
├───────┼───────┼─────────┼─────────────────┤
│     1 │ 1     │ (1,'1') │               1 │
├───────┼───────┼─────────┼─────────────────┤
│     2 │ 2     │ (2,'2') │               2 │
├───────┼───────┼─────────┼─────────────────┤
│     3 │ 3     │ (3,'3') │            ᴺᵁᴸᴸ │
├───────┼───────┼─────────┼─────────────────┤
│     4 │ 4     │ (4,'4') │               1 │
└───────┴───────┴─────────┴─────────────────┘
┏━━━━━━━┳━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ hello ┃ world ┃ tuple   ┃ sometimes_nulls ┃
┡━━━━━━━╇━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│     5 │ 5     │ (5,'5') │               2 │
├───────┼───────┼─────────┼─────────────────┤
│     6 │ 6     │ (6,'6') │            ᴺᵁᴸᴸ │
├───────┼───────┼─────────┼─────────────────┤
│     7 │ 7     │ (7,'7') │               1 │
├───────┼───────┼─────────┼─────────────────┤
│     8 │ 8     │ (8,'8') │               2 │
├───────┼───────┼─────────┼─────────────────┤
│     9 │ 9     │ (9,'9') │            ᴺᵁᴸᴸ │
└───────┴───────┴─────────┴─────────────────┘
```

and running `SELECT number AS hello, toString(number) AS world, (hello, world) AS tuple, nullIf(hello % 3, 0) AS sometimes_nulls FROM system.numbers LIMIT 10 SETTINGS max_block_size = 5 FORMAT PrettyNoEscapes;`

```
┏━━━━━━━┳━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ hello ┃ world ┃ tuple   ┃ sometimes_nulls ┃
┡━━━━━━━╇━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│     0 │ 0     │ (0,'0') │            ᴺᵁᴸᴸ │
├───────┼───────┼─────────┼─────────────────┤
│     1 │ 1     │ (1,'1') │               1 │
├───────┼───────┼─────────┼─────────────────┤
│     2 │ 2     │ (2,'2') │               2 │
├───────┼───────┼─────────┼─────────────────┤
│     3 │ 3     │ (3,'3') │            ᴺᵁᴸᴸ │
├───────┼───────┼─────────┼─────────────────┤
│     4 │ 4     │ (4,'4') │               1 │
└───────┴───────┴─────────┴─────────────────┘
┏━━━━━━━┳━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ hello ┃ world ┃ tuple   ┃ sometimes_nulls ┃
┡━━━━━━━╇━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│     5 │ 5     │ (5,'5') │               2 │
├───────┼───────┼─────────┼─────────────────┤
│     6 │ 6     │ (6,'6') │            ᴺᵁᴸᴸ │
├───────┼───────┼─────────┼─────────────────┤
│     7 │ 7     │ (7,'7') │               1 │
├───────┼───────┼─────────┼─────────────────┤
│     8 │ 8     │ (8,'8') │               2 │
├───────┼───────┼─────────┼─────────────────┤
│     9 │ 9     │ (9,'9') │            ᴺᵁᴸᴸ │
└───────┴───────┴─────────┴─────────────────┘
```

So this is the behaviour I observe even when I can see with lldb that for the first query `color = true` and in the second query `color = false`.

Avogar commented:
The tests are failing in the CI because when we generate reference files we redirect the output to the file, so don't write to the terminal and our `auto` mode disables escape sequences. And this is ok.
I just downloaded the binary from fasttest with your cahnges and when I run command `clickhouse client -q "select 42 as value format Pretty"` without redirection to the file, it prints escape sequences, when I run `clickhouse client -q "select 42 as value format Pretty" > reference`, then it doesn't pring escape sequences. So, seems like everything is ok.

Blargian committed:
#58363 - add change to SettingsChangesHistory.h and fix failing tests which use pretty by setting output_format_pretty_color=1

Avogar commented on src/Core/SettingsChangesHistory.h:
```diff
@@ -85,6 +85,7 @@ static std::map<ClickHouseVersion, SettingsChangesHistory::SettingsChanges> sett
               {"input_format_parquet_allow_missing_columns", false, true, "Allow missing columns in Parquet files by default"},
               {"input_format_orc_allow_missing_columns", false, true, "Allow missing columns in ORC files by default"},
               {"input_format_arrow_allow_missing_columns", false, true, "Allow missing columns in Arrow files by default"}}},
+              {"output_format_pretty_color",true,"auto","Setting is changed to allow also for auto value, disabling ANSI escapes if output is not a tty"}
```
```suggestion
              {"input_format_arrow_allow_missing_columns", false, true, "Allow missing columns in Arrow files by default"},
              {"output_format_pretty_color", true, "auto", "Setting is changed to allow also for auto value, disabling ANSI escapes if output is not a tty"}}},
```

Blargian committed:
Update src/Core/SettingsChangesHistory.h

Co-authored-by: Kruglov Pavel <48961922+Avogar@users.noreply.github.com>

Blargian committed:
fix failing 02421 stateless and (hopefully) fix failing integration tests

Blargian committed:
Merge remote-tracking branch 'refs/remotes/origin/58363_disable_ansi_pretty_automatically' into 58363_disable_ansi_pretty_automatically

Blargian commented:
@Avogar sorry for the delay, I got stuck now on trying to fix the last remaining tests which are failing. I don't find any test `test_parallel_replicas_custom_key_failover` in my local repository, only `test_parallel_replicas_custom_key`, but I don't see anything related to Pretty format there. As for Flaky tests I don't even know where to begin to fix that...

Avogar commented:
> I got stuck now on trying to fix the last remaining tests which are failing

Actually, all failed tests are not related to the changes in this PR. `test_parallel_replicas_custom_key_failover` test is broken in master and our team is looking into it. Integration flaky check failed due to timeout and it's an infrastracture problem. Stateless tests flaky check failed because of test `01074_partial_revokes`, but it's expected for this test, it should not be run in parallel (we need to add special tag to this test to avoid it).

Blargian commented:
@Avogar Any other open issues around this part of the codebase for Pretty/Formats you can recommend for me to work on?

alexey-milovidov commented:
Amazing! It works:

```
watch -n1 'clickhouse-local --query "SELECT now() FORMAT Pretty"'
```
